circuit_maintenance_parser/utils.py

```diff
@@ -1,27 +1,116 @@
+        )
+        return location_coordinates
+
+    def get_location_from_local_file(self, city: str) -> Tuple[float, float]:
+        """Get location from Local DB."""
+        city_name = city.split(", ")[0]
+        country = city.split(", ")[-1]
+
+        if self.db_location is not None:
+            # 1. We try to match city name and country
+            location = self.db_location.query(f"city_ascii == '{city_name}' & country == '{country}'")
+            if not location.empty:
+                logger.debug("Used city name and country to resolve %s to %s from local locations DB.", city, location)
+                return (location.lat.values[0], location.lng.values[0])
+
+            # 2. We try to match only city name
+            location = self.db_location.query(f"city_ascii == '{city_name}'")
+            if not location.empty:
+                logger.debug("Used only city name to resolve %s to %s from local locations DB.", city, location)
+                return (location.lat.values[0], location.lng.values[0])
+
+        logger.debug("City %s was not resolvable in the local locations DB.", city)
+        raise ValueError
+
+    @staticmethod
+    @backoff.on_exception(
+        backoff.expo, (GeocoderUnavailable, GeocoderTimedOut, GeocoderServiceError), max_time=10, logger=logger,
+    )
+    def get_location_from_api(city: str) -> Tuple[float, float]:
+        """Get location from API."""
         geolocator = Nominatim(user_agent="circuit_maintenance")
         location = geolocator.geocode(city)  # API call to OpenStreetMap web service
-        timezone = (
-            tzwhere.tzwhere()
-        )  # TODO: Offline loading of timezone location data is quite slow. Look for better alternative
-        return timezone.tzNameAt(location.latitude, location.longitude)
-    except (GeocoderUnavailable, GeocoderTimedOut, GeocoderServiceError):
-        raise ParserError(  # pylint: disable=raise-missing-from
-            "Cannot connect to the remote Geolocator API to determine timezone"
-        )
+        logger.debug("Resolved %s to %s from OpenStreetMap webservice.", city, location)
+        return (location.latitude, location.longitude)
+
+    def city_timezone(self, city: str) -> str:
+        """Get the timezone for a given city.
+
+        Args:
+            city (str): Geographic location name
+        """
+        if self.timezone is not None:
+            try:
+                latitude, longitude = self.get_location(city)
+                timezone = self.timezone.tzNameAt(latitude, longitude)
+                if not timezone:
+                    # If even with the location we can't find the timezone, we try getting the location
+                    # from API.
+                    latitude, longitude = self.get_location_from_api(city)
+                    timezone = self.timezone.tzNameAt(latitude, longitude)
```

Comment:
Shouldn't be needed since `get_location` already falls through on its own to calling `get_location_from_api` if `get_location_from_local_file` fails?

Reply:
in theory yes, but I found a case, with the original DB data, that was resolving to Sydney coordinates and the timezone resolution was returning `None`, maybe due some bug on that library, because with the coordinates from the API was working well (some decimals of difference)